File: artifacts/api-server/src/lib/retrieval.ts
import { sql } from "drizzle-orm";
import { db, documentChunks } from "./db";
import { embedOne, toVectorLiteral } from "./embeddings";

export type RetrievedChunk = { content: string; score: number; documentId: number };

// Hybrid search: vector cosine (if embedding available) + lexical ILIKE.
// Falls back to lexical-only when embeddings are unavailable.
export async function hybridSearch(
  tenantId: number,
  query: string,
  limit = 5,
): Promise<RetrievedChunk[]> {
  const trimmed = query.trim();
  if (!trimmed) return [];

  const vec = await embedOne(trimmed);

  const rows = await db.execute(sql`
    WITH lexical AS (
      SELECT id, content, document_id,
        ts_rank_cd(to_tsvector('simple', content), plainto_tsquery('simple', ${trimmed})) AS lex_score
      FROM document_chunks
      WHERE tenant_id = ${tenantId}
        AND content ILIKE ${"%" + trimmed + "%"}
      LIMIT 20
    ),
    vector_match AS (
      ${
        vec
          ? sql`SELECT id, content, document_id,
                1 - (embedding <=> ${toVectorLiteral(vec)}::vector) AS vec_score
            FROM document_chunks
            WHERE tenant_id = ${tenantId}
              AND embedding IS NOT NULL
            ORDER BY embedding <=> ${toVectorLiteral(vec)}::vector
            LIMIT 20`
          : sql`SELECT id, content, document_id, 0::float AS vec_score
            FROM document_chunks WHERE false`
      }
    ),
    merged AS (
      SELECT
        COALESCE(v.id, l.id) AS id,
        COALESCE(v.content, l.content) AS content,
        COALESCE(v.document_id, l.document_id) AS document_id,
        COALESCE(v.vec_score, 0) AS vec_score,
        COALESCE(l.lex_score, 0) AS lex_score
      FROM vector_match v
      FULL OUTER JOIN lexical l ON l.id = v.id
    )
    SELECT id, content, document_id,
           (0.7 * vec_score + 0.3 * LEAST(lex_score, 1.0)) AS score
    FROM merged
    ORDER BY score DESC
    LIMIT ${limit}
  `);

  // drizzle node-postgres returns { rows }
  const list = (rows as unknown as { rows: Array<{ content: string; score: number; document_id: number }> })
    .rows;
  return list.map((r) => ({
    content: r.content,
    score: Number(r.score) || 0,
    documentId: r.document_id,
  }));
}

// Best-effort cosine confidence — returns 0 when there's nothing relevant.
export function confidenceFrom(chunks: RetrievedChunk[]): number {
  if (chunks.length === 0) return 0;
  const top = chunks[0]?.score ?? 0;
  return Math.max(0, Math.min(1, top));
}

File: artifacts/api-server/app/services/retrieval.py
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession
from app.services.embeddings import embed_one


async def hybrid_search(db: AsyncSession, tenant_id: int, query: str, limit: int = 5) -> list[dict]:
    embedding = await embed_one(query)

    if embedding is None:
        # Lexical-only fallback
        sql = text("""
            SELECT dc.id, dc.content,
                   ts_rank_cd(to_tsvector('simple', dc.content), plainto_tsquery('simple', :q)) AS score
            FROM document_chunks dc
            JOIN documents d ON d.id = dc.document_id
            WHERE dc.tenant_id = :tid
              AND d.status = 'indexed'
              AND dc.content ILIKE '%' || :q || '%'
            ORDER BY score DESC
            LIMIT :lim
        """)
        result = await db.execute(sql, {"tid": tenant_id, "q": query, "lim": limit})
        rows = result.mappings().all()
        return [{"content": r["content"], "score": float(r["score"]), "document_id": 0} for r in rows]

    vec_str = "[" + ",".join(str(x) for x in embedding) + "]"

    sql = text("""
        WITH vec AS (
            SELECT dc.id,
                   dc.content,
                   dc.document_id,
                   1 - (dc.embedding <=> :vec::vector) AS vec_score
            FROM document_chunks dc
            JOIN documents d ON d.id = dc.document_id
            WHERE dc.tenant_id = :tid
              AND d.status = 'indexed'
              AND dc.embedding IS NOT NULL
            ORDER BY dc.embedding <=> :vec::vector
            LIMIT 20
        ),
        lex AS (
            SELECT dc.id,
                   ts_rank_cd(to_tsvector('simple', dc.content), plainto_tsquery('simple', :q)) AS lex_score
            FROM document_chunks dc
            JOIN documents d ON d.id = dc.document_id
            WHERE dc.tenant_id = :tid
              AND d.status = 'indexed'
              AND dc.content ILIKE '%' || :q || '%'
            LIMIT 20
        )
        SELECT
            COALESCE(vec.id, lex.id) AS id,
            COALESCE(vec.content, (
                SELECT content FROM document_chunks WHERE id = lex.id
            )) AS content,
            COALESCE(vec.document_id, 0) AS document_id,
            (COALESCE(vec.vec_score, 0) * 0.7 + COALESCE(lex.lex_score, 0) * 0.3) AS score
        FROM vec
        FULL OUTER JOIN lex ON vec.id = lex.id
        ORDER BY score DESC
        LIMIT :lim
    """)
    result = await db.execute(sql, {"tid": tenant_id, "vec": vec_str, "q": query, "lim": limit})
    rows = result.mappings().all()
    return [
        {"content": r["content"], "score": float(r["score"]), "document_id": int(r["document_id"])}
        for r in rows if r["content"]
    ]
